GHzDACs/jump_table.py

```diff
@@ -0,0 +1,629 @@
+        """
+
+        :param start_addr: start address
+        :param list[JumpEntry] jumps: the jump table entries
+        :param list[int] counters: the counter values
+        """
+        self._startAddr = 0
+        self.counters = self._initialize_counters(counters=counters)
+        self.start_addr = start_addr
+        self.jumps = jumps
+
+    @classmethod
+    def _initialize_counters(cls, counters=None):
+        if counters is None:
+            c = [0, 0, 0, 0]
+        elif len(counters) > cls.NUM_COUNTERS:
+            raise ValueError("Cannot have more than 4 counters.")
+        else:
+            if any([x > cls.COUNT_MAX for x in counters]):
+                raise ValueError(
+                    "Counter values must fit in {} bits.".format(
+                        cls.COUNTER_BITS
+                    )
+                )
+            c = list(counters)
+            while len(c) < cls.NUM_COUNTERS:
+                c.append(0)
+        return c
+
+    def __str__(self):
+        counter = '\n'.join(
+            "Counter {}: {}".format(
+                i, c
+            ) for i, c in enumerate(self.counters)
+        )
+        start_addr = 'Start address: %d' % self.start_addr
+        jump = '\n'.join(
+            "-JUMP ENTRY {}-\n{}".format(
+                i, str(jump)
+            ) for i, jump in enumerate(self.jumps)
+        )
+        return '\n'.join([counter, start_addr, jump])
+
+    def toString(self):
+        """Serialize jump table to a byte string for the FPGA"""
+        data = np.zeros(self.PACKET_LEN, dtype='<u1')
+        # Set counter values. Each one is 4 bytes
+        for i, c in enumerate(self.counters):
+            data[i * 4:(i + 1) * 4] = littleEndian(c, 4)
+        # Set start address
+        data[16:19] = littleEndian(self.start_addr, 3)
+        data[19:22] = littleEndian(self.start_addr, 3)
+        # Start op code
+        data[22] = 5
+        data[23] = 0
+        for i, jump in enumerate(self.jumps):
+            ofs = 24 + i * 8
+            data[ofs:ofs+8] = jump.as_bytes()
+        return data.tostring()
+
+    def pretty_string(self):
+        s = ''
+        data = self.toString()
+        i = 0
+        # counters
+        for j in range(4):
+            s += '{0:02x} {1:02x} {2:02x} {3:02x}'.format(*[ord(x) for x in data[i:i+4]])
+            i += 4
+            s += '\n'
+        # table commands
+        for j in range((528-16)//8):
+            s += '{0:02x} {1:02x} {2:02x}  '.format(*[ord(x) for x in data[i:i+3]])
+            i += 3
+            s += '{0:02x} {1:02x} {2:02x}  '.format(*[ord(x) for x in data[i:i+3]])
+            i += 3
+            s += '{0:02x} {1:02x}  '.format(*[ord(x) for x in data[i:i+2]])
+            i += 2
+            s += '\n'
+        # print i
+        # print len(data)
+        assert(i == len(data))
+        return s
+
+
+# TEST FUNCTIONS
+
+def testNormal(stopAddr):
+    """Test END function
```

Comment:
I think the do cstring here contains errors. Offsets are not handled, so I think this function assumes the calling code is dealing with the offsets (Which is great because that separates the serialization in this module from offset handling which should be done by the individual build classes in `dac.py`). Therefore, I think passing a `stopAddr` of 5 would actually cause the system to stop after the second pulse. This mistake probably exists in the rest of the test functions too. @pomalley can we sit together and fix this?